### ISSUE TITLE
Add Code Folding

### ISSUE THREAD

**Issue by waterlubber** [helix-editor/helix#1840]
As far as I can tell, Helix does not (yet) support [code folding](https://en.wikipedia.org/wiki/Code_folding). 
Syntax or LSP/Treesitter based code folding is probably my most missed feature from nvim, 

It would probably be best to add it in a similar manner to [nvim's approach](https://neovim.io/doc/user/fold.html), although initial patches might only support syntax-based folding if it's simpler. 

There's a chance I might have missed this feature - I couldn't find it in the editor, documentation, or as an issue, so I figured I'd make the feature request. Apologies if I have missed something, though!

**Comment by i3elj** [comment]
Just for the matter of visualization:

![foldingInNeovim](https://user-images.githubusercontent.com/61887184/169597023-c7f333b7-abbc-41d3-90af-7719c5813ef9.png)

I think this is a pretty nice UI

**Comment by naveedpash** [comment]
any progress on this?

**Comment by EriKWDev** [comment]
I find myself really missing this. In rust when you have a bunch of "pub mod foo { /* ... */ }" it can be very handy to be able to fold them away while not working on them.

**Comment by crispinb** [comment]
I find emacs-style narrow (aka hoist) more useful than the (more common) fold. Folding complects navigation with focusing, but navigation is already covered by lsp. Narrowing is extremely useful for focusing singly on a chunk of code to be worked on.

**Comment by Zoybean** [comment]
I've never encountered narrowing before, but from googling it just now, it seems like it might serve a different use case than folding, and maybe they could both be added?

**Comment by crispinb** [comment]
Sure, they're not incompatible. Emacs has both. 

My perspective (no more than that!) is that folding doesn't offer much given its relative UX complexity (multiple operations and keybindings). It does 2-3 things: offers a code overview + helps with navigation, and allows you to focus on one section of code by folding  others. 

To me, navigation is already taken care of by many existing helix/lsp features. Focus is managed much more simply by narrowing - it's just a single and simple pair of operations (narrow/expand), which fit like a glove with a central existing concept (the selection - to which you can narrow).

That leaves potentially the code overview as folding's one unique use. Seems like a sledgehammer to crack a nut to me, and probably handled better by other (and again simpler) means, but obviously it's pretty subjective. Folding is far more widely implemented than is hoisting, so I guess people must find it useful.

**Comment by Zoybean** [comment]
Yeah, I admit that from my experience with folding in vim, it does feel clunky even when it's working properly, so I see some value in investigating other possibilities for a code overview feature aside from traditional folding.

**Comment by s0LA1337** [comment]
> Just for the matter of visualization:
> 
> ![foldingInNeovim](https://user-images.githubusercontent.com/61887184/169597023-c7f333b7-abbc-41d3-90af-7719c5813ef9.png)
> 
> I think this is a pretty nice UI (it's from neovim)

this is only possible once virtual text is implemented.

I believe this issue (to have it done in a imo nice way) would be to wait once #417 is merged.

**Comment by Itay123TheKing** [comment]
I don't think it makes sense to drop folds in favour of narrowing, but I'd be glad to have narrowing in core. I can imagine it in it's own submenu with equivalents for all tree-sitter movement commands (`A-[iopn]`), and exiting and entering narrow. If it does get in though, I would like to see it implemented as in https://github.com/Malabarba/fancy-narrow, though at narrowing may be deservant of its own issue. 

Do note I have 0 experience with narrowing and I'm saying all of this only after a tiny bit of research into what it even is and why its useful. I think done like in https://github.com/Malabarba/fancy-narrow it would get a very real usecase of helping with focusing on code and let the other commands I've suggested above operate much more smoothly, unlike when just creating a virtual buffer.

**Comment by Eloitor** [comment]
For python it would be useful to be able to fold docstrings

**Comment by pashadia** [comment]
> For python it would be useful to be able to fold docstrings

I would do exactly the opposite: Fold everything BUT the docstrings

**Comment by frondeus** [comment]
So, #411 is closed now, which I guess was a blocker.

**Comment by SoraTenshi** [comment]
> How would this be implemented in codebase?

i would assume that we would also create tree-sitter queries for this, and then basically manipulate the text rendering in a way to skip all the lines that are within that context.
Also; rendering the "context core information" such as function name, etc. as virtual text.

**Comment by pascalkuthe** [comment]
A conceal API was excluded from #5420. I have a prototype for that locally but it needs more work. I am doing some refactors of the vritual text API anyway atm but I probably won't include the conceal API with that as it's a bit more complex.

In particular we must be careful that vertical/horitzontal movement moves past the concealed text (vrtical movement already does) while any other movement/search etc. that would place a selection inside the concealed text expands the concealed text. This requires centrelazing the way we set selctions which is something we want to do anyway.

**Comment by MrGibus** [comment]
> we must be careful that vertical/horitzontal movement moves past the concealed text (vrtical movement already does) while any other movement/search etc.

I think this is a very important aspect. The vim implementation in vscode had a very annoying flaw in that moving vertically past a fold opened it, as well as several other issues related to it, which probably still plague it to this day.

**Comment by Mateiadrielrafael** [comment]
Are there plans to also support marker based folding? That's a pretty important part of my workflow.

**Comment by 87113** [comment]
Hi there, another use case for folding (vs narrowing) is for markdown. 

I like to be able to fold titles or list for example. 
Like that I can see the file's structure without all the text around. 
It's a bit like an inside TOC.

**Comment by Connortsui20** [comment]
Is there any plan to work on this? There seems to be a lot of additional features on top of basic code folding that would be interesting to have, but there is no basic implementation yet. 

From #5798 it seems like a conceal API will not be fully implemented soon, but I'm not sure it would be absolutely necessary for basic code folding.

The way VSCode does folding is simply adding dots to the end of the first line, darkening the line, and removing the folded lines from sight:

![image](https://github.com/helix-editor/helix/assets/87130162/53c214da-c759-468a-bbce-967dac67caf5)
![image](https://github.com/helix-editor/helix/assets/87130162/95d3c14a-0463-471e-987e-b85c7818898a)

Maybe this would be a good first step?

**Comment by pascalkuthe** [comment]
Code folding and conceal need the same lower level API being able to hide part of the document and replace it with some virtual text.

That API doesn't exist yet and is not easy to implement. When it exists implementing code folding (and conceal) will not be that hard. The hardest parts are already done in #5420 but further work is needed and I don't have the time and motivation to finish that right now

**Comment by frankleigh123** [comment]
Usually I only want a list of functions in the current displayed buffer. I use vsplit-new and then :pipe grep.... A quicker way to achieve this would be useful.

**Comment by eliliam** [comment]
A basic version of folding would be amazing, even if it is behind a flag that is disabled by default until it's fully complete. Is there a rough time frame on when work will start on this?

**Comment by AE86Trueno** [comment]
Reaching here when trying to find how to fold in Helix.
Fold is one of the Vim function I use all the time. In my use case, I have a text file (currently already 25k lines of text). I break it into section and sub-sections and require Fold for me to navigate.

**Comment by GreatC0der** [comment]
Really waiting for these feature.

**Comment by Eloitor** [comment]
This would be useful for saturn notebook support: https://github.com/mrzv/saturn

**Comment by camilodlt** [comment]
Hello, I've been using Helix for a week and it's amazing. Thanks! . What about folding with the "region" command.

It's done on vscode and some lsps recognize them apparently: 

` 
# Unfolded code

# region set seeds
random.seed(1)
# endregion

# Folded : 

# region set seeds ...
` 

Of course, different languages have different markers for comments [vs code regions](https://code.visualstudio.com/docs/editor/codebasics#_folding)

**Comment by Mateiadrielrafael** [comment]
@camilodlt In vim I use `# {{{` and `# }}}` comments for folding (this is referred to as "marker based folding" I think), so the nicest option would be having configurable markers (so both our comments would fold as intended)

**Comment by tgross35** [comment]
As far as the logic for how to fold: nvim uses `folds.scm`, with an `@fold` capture. There are already some of these in the codebase, e.g. https://github.com/helix-editor/helix/blob/7e389b67c24dfe4466112c988b240c807e7e2414/runtime/queries/latex/folds.scm

**Comment by luccahuguet** [comment]
> Code folding and conceal need the same lower level API being able to hide part of the document and replace it with some virtual text.
> 
> That API doesn't exist yet and is not easy to implement. When it exists implementing code folding (and conceal) will not be that hard. The hardest parts are already done in #5420 but further work is needed and I don't have the time and motivation to finish that right now

Hi @pascalkuthe, is there a issue to track this lower level API that you mention, and if not, can I create one?

**Comment by dawkrish** [comment]
so are there any updates on code-folding ?

**Comment by saifulapm** [comment]
Any update on code folding ?

**Comment by pashadia** [comment]
> I really like this feature and use a lot in other editors but in helix I "indirectly replace" it by using more the symbol picker.

While I agree that using a symbol picker is preferable _when possible_, when working with poorly maintained codebases, you sometimes encounter functions spanning hundreds of lines, and several distinct flows.

**Comment by valpackett** [comment]
> using a symbol picker is preferable when possible

Not everyone agrees: [A Missing IDE Feature: Fold Method Bodies by Default](https://matklad.github.io/2024/10/14/missing-ide-feature.html)

**Comment by lukasjuhrich** [comment]
Piggybacking on @valpackett's comment: I'd also warn against the blanket judgement "if code folding is necessary, then it must be due to a poor quality codebase or it can be mediated by a symbol picker": There can be various other reasons. For instance because i might be looking at a `compose.yml` with large environment configuration sections which don't always matter, but take up half a screen of space.
Similar things hold for JSON or XML (popular for msbuild project files). Sometimes, symbol navigation might not even exist or be reasonable to use.

Helix does much more than just allow you to write rust code, thus it should aid the user in a variety of environments. Code folding would help a lot with that.

**Comment by emilyyyylime** [comment]
Not to mention a batteries-included editor _should_ aim to improve the experience of editing bad code. A lot of actively maintained code out there reeks of technical debt and Helix can't just turn its back and pretend it's someone else's problem. As mentioned in the blog a text-mode function view has many non trivial advantages like being able to search within a selection, multi-cursor rename, wrap in an inline module, or freely move (cut+paste) functions en masse

**Comment by hoichi** [comment]
> As mentioned in the blog a text-mode function view has many non trivial advantages like being able to search within a selection, multi-cursor rename, wrap in an inline module, or freely move (cut+paste) functions en masse

This. FWIW, out of modern terminal/TUI tools, I love flexibility of [broot](https://dystroy.org/broot/) that blurs the line between terminal and TUI. For a clever, terminal-friendly, post-modern editor, I think we could use more of that.

Also, any kind of filtering/narrowing in the same buffer could use the screen estate more effectively than most IDEs. And as for things like multi-cursor rename (provided it can’t be done via LSP, of course), I guess you could integrate some of it in a separate [outline] mode, but maybe it could be done seamlessly and consistently in the same one.

**Comment by TheZoq2** [comment]
> I really like this feature and use a lot in other editors but in helix I "indirectly replace" it by using more the symbol picker.

I feel like people make the same argument against a file picker, and it misses some points. Fuzzy matching is incredible for finding an exact file, or an exact function *if you know what to look for*. But sometimes you don't know exactly what to look for, sometimes because the code base is bad, but often because it is big and you don't have enough experience in it to know exactly what is in there. Being able to "casually" scroll through the things then is super useful

**Comment by pascalkuthe** [comment]
The discussion here is not fruitful. We will add folding in the future but it's a non-trivial task that requires changing our core positioning code. Some of the groundwork was laid in #6417  so this is definitely something I *will* do and *is* possible but it's a complicated feature to implement. It can likely only be done by me (or somebody very experienced/another maintainer in close collaboration with me). Other things are taking priority, there is only so much bandwidth in a volunteer-based project.

I will leave this issue unlocked for now in case there is some *implementation related* discussion but I will mark these sort of meta discussions (including comments asking for ETA) as off-topic. You can use a github discussion or matrix  (or other platforms) for that sort of discussion but an issue is not the right place for that

**Comment by n8himmel** [comment]
I  miss code folding so much.
Just to keep the thread here alive.   
-- Thanks for an amazing editor!

**Comment by eyal-di** [comment]
why is this so hard to implement?(honest question)
god i'm this close to going back to VScode and i dont want to! :(
considering leaning rust just for this lol

**Comment by jbethune** [comment]
> why is this so hard to implement?(honest question)

Scroll up to https://github.com/helix-editor/helix/issues/1840#issuecomment-2427603088

**Comment by amandalewis1231-lgtm** [comment]
Is this specific to a certain os/browser

**Comment by ikseski** [comment]
Helix is so good that I can use it even without code folding, although I admit there are high hopes to get it finally shipped in the community :)

**Comment by Blankeos** [comment]
Hi all, if you can't wait for code folding like I do... I made this minor fork of Helix for the meantime that I'll make sure to keep up-to-date to upstream as much as possible. Mostly inspired from Zed (since I come from there, and it's a crucial part of my flow). Install instructions are [here](https://github.com/Blankeos/helix/pull/1). Lots to polish but I'll be dogfooding it until it's ready.

https://github.com/user-attachments/assets/6e87b96e-d1a3-485a-957d-b45619eba355
